### DISCUSSION TITLE
6.0 planning

### DISCUSSION THREAD

**Discussion by edsrzf** [vimeo/psalm#10701] [Psalm development (AKA internals)]
I'd be keen to see a bit more of a plan for the 6.0 release. I've only seen comments and PRs here and there that I've pieced together.

Here are the major changes I'm aware of:

* Requiring PHP 8.1+ to run Psalm. (#10024)
* Updating `amphp/amp` to 3.0. (#10024)
* Updating `nikic/php-parser` to 5.0. (#10567)
* A few BC breaks mentioned in [UPGRADING.md](https://github.com/vimeo/psalm/blob/master/UPGRADING.md)

Possible other things to consider:

* Drop support for analyzing PHP versions below 7. `nikic/php-parser` 5.0 has dropped the dedicated parser for PHP 5.x. See its [upgrade notes](https://github.com/nikic/PHP-Parser/blob/master/UPGRADE-5.0.md#php-5-parsing-support). However maybe it still works well enough that it's worth keeping Psalm analysis. 🤷 I actually can't see any places Psalm checks for versions lower than 7.
* Type pool? #9328 Or at least the necessary incompatible changes to make a type pool possible.
* Adopt more PHP 8.0/8.1 features (eg promoted constructor properties)

What do we want to be sure to get done before a release? Is there anything else? Is there a timeline to target (aspirational or otherwise)?

**Comment by danog** [comment]
Personally, the only blockers I see are:

* https://github.com/vimeo/psalm/pull/10577
* https://github.com/vimeo/psalm/pull/9778

I'm quite busy these days, but I will try to get some work done on those PRs ASAP.

**Reply by edsrzf** [reply]
No worries, I understand folks are busy. I'm not in any particular rush. I just want to understand what peoples' plans were and look for opportunities to help. Hopefully others will find it useful too.

**Comment by Jean85** [comment]
Please keep in mind that since only 6.0 will include #10567, this means that:
 * Psalm 6 is a blocker to upgrade to PHPUnit 11
 * ...and viceversa, so...
 * you will need to upgrade to PHPUnit 11 in the same step of upgrading to Psalm

**Reply by weirdan** [reply]
We run PHPUnit 9 on master (e.g https://github.com/vimeo/psalm/actions/runs/8041668107/job/21961230131#step:11:13), so at least you don't have to upgrade PHPUnit when you upgrade Psalm.

**Reply by SCIF** [reply]
@Jean85 , could you please elaborate? dev-dependencies of psalm or its deps has nothing to do with installation of psalm into your project.

**Reply by Jean85** [reply]
@SCIF and in fact it's an issue of non dev dependencies.

nikic/parser is a non dev requirement of Psalm, but also a (non direct) non dev dependency of PHPUnit. The issue arises in the fact that PHPUnit 11 require nikic/parser 5, while Psalm requires 4.16, so if I'm an end user of Psalm, I can't upgrade to PHPUnit 11 due to this shared dependency.

**Reply by SCIF** [reply]
@Jean85 , in case if you are a user of Psalm then you don't need to install dev-deps of Psalm. The same applies to nikic/parser.

**Reply by Jean85** [reply]
@SCIF I need PHPUnit for myself, not because it's a dev dependency of Psalm. The only way around that I have is the PHAR, or creating a separate composer.json.

**Reply by Jean85** [reply]
A simple reproducer is:
```
$ composer init
$ composer require --dev vimeo/psalm
[...]
Using version ^5.22 for vimeo/psalm
$ composer require --dev phpunit/phpunit ^11 -W
```
This gets you:
```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - phpunit/php-code-coverage[11.0.0, ..., 11.0.1] require nikic/php-parser ^5.0 -> found nikic/php-parser[v5.0.0, v5.0.1] but these were not loaded, likely because it conflicts with another require.
    - phpunit/phpunit[11.0.0, ..., 11.0.4] require phpunit/php-code-coverage ^11.0 -> satisfiable by phpunit/php-code-coverage[11.0.0, 11.0.1].
    - Root composer.json requires phpunit/phpunit ^11 -> satisfiable by phpunit/phpunit[11.0.0, ..., 11.0.4].
```

**Reply by edsrzf** [reply]
Looks like you're correct that there is a conflict with PHPUnit 11. With that version, `phpunit/php-code-coverage` has dropped support for `nikic/php-parser` v4.

You'd need to either stick with PHPUnit 10 for now (which is still supported) or use Psalm from `dev-master`.

**Reply by SCIF** [reply]
Thanks, @Jean85 , sorry, I thought the problem is other way round — php-parser requires phpunit 7-9 and that's the issue, but the problem in code-coverage conflicting with php-parser. Clear now.

**Comment by SCIF** [comment]
Guys, any thoughts on the v6 release? Whether `master` is v6 branch?

Master is being diverged for a long time and the difference is quite significant. I heavily appreciate all contributions to Psalm and advocate using Psalm a lot. But it looks like the complexity of a contribution to Psalm is sky-high and I worry that the project is slowing down its development. I tried to think about it and came to the following conclusions:
1. A low-hanging fruit is: to eliminate the diverseness of `master` and current version branches. If `master` is a next-gen branch then state it in the contribution guide and make the stable branch mergable to `master`.
2. Put efforts into simplification/refactoring tasks rather than introduction of new features. What I found during the development of my contributions — was a lot of copy-pasted code which is a signal to think about refactoring.

One more time — I didn't want to blame/insult/push anybody but tried to initiate a discussion about the future of this great tool.

**Reply by jorgsowa** [reply]
I have the same feeling and it's upsetting me. There are many nice features on master and it would be awesome to release it already. For some of the projects, the current version of the psalm is also a blocker for PHPUnit 11.

Are there any things that the community is able to help with? I could allocate some hours myself but without a general direction/timeline/plan it's impossible to know what's expected.

**Comment by carlos-granados** [comment]
Now that PHPStan 2.0 and Rector 2.0 are both available, both supporting PHP Parser v5 exclusively, it is important to get this Psalm v6 version released. Otherwise it may be difficult for some tools which depend on PHP Parser to support both sets of tools.
